### ISSUE TITLE
Questions about how the gradient is backpropagated

### ISSUE THREAD

**Issue by GaoLianger** [HuXiaoling/TopoLoss#29]
Thank you for sharing the code. I have a conceptual question about optimizing a loss function in NumPy without using automatic differentiation (e.g., PyTorch's requires_grad). How to manually compute gradients of a loss function in NumPy (without automatic differentiation like requires_grad) and optimize parameters using gradient descent?

**Comment by Wangxin-11** [comment]
> Thank you for sharing the code. I have a conceptual question about optimizing a loss function in NumPy without using automatic differentiation (e.g., PyTorch's requires_grad). How to manually compute gradients of a loss function in NumPy (without automatic differentiation like requires_grad) and optimize parameters using gradient descent?
May I ask if you have solved this problem?

**Comment by GaoLianger** [comment]
> > Thank you for sharing the code. I have a conceptual question about optimizing a loss function in NumPy without using automatic differentiation (e.g., PyTorch's requires_grad). How to manually compute gradients of a loss function in NumPy (without automatic differentiation like requires_grad) and optimize parameters using gradient descent?
> > May I ask if you have solved this problem?
Thank you for your attention. I haven't solved this problem yet. It's mainly because I'm not familiar with the approach of gradient propagation in NumPy. I feel that I need to customize the corresponding backward function, but I haven't found it in the code.
